editoast/src/views/timetable/track_occupancy_utils.rs

```diff
@@ -0,0 +1,269 @@
+                    if operational_point.0 == operational_point_id {
+                        return Some(&path_item.id.0);
+                    }
+                }
+                OperationalPointIdentifier::OperationalPointDescription { trigram, .. } => {
+                    let operational_points = path_item_cache
+                        .get_from_trigram(&trigram.0)
+                        .expect("The operational points are supposed to exist");
+                    if operational_points
+                        .iter()
+                        .any(|op| op.obj_id == operational_point_id)
+                    {
+                        return Some(&path_item.id.0);
+                    }
+                }
+                OperationalPointIdentifier::OperationalPointUic { uic, .. } => {
+                    let ops = path_item_cache
+                        .get_from_uic(*uic)
+                        .expect("The operational points are supposed to exist");
+                    if ops.iter().any(|op| op.obj_id == operational_point_id) {
+                        return Some(&path_item.id.0);
+                    }
+                }
+            }
+        }
+    }
+    None
+}
+
+/// Extract track ranges and path item positions from pathfinding result
+/// Returns None if pathfinding failed
+///
+/// Note: The timetable can have a pathfinding failure for a train schedule
+/// because the train schedule is not valid. We skip it because we don't want
+/// to compute the track occupancy for an invalid train schedule.
+/// It's not a problem because we are just looking for the track occupancy
+/// of the operational point for the valid trains.
+pub fn extract_pathfinding_data(
+    pathfinding: &PathfindingResult,
+) -> Option<(&Vec<core_client::pathfinding::TrackRange>, &Vec<u64>)> {
+    match pathfinding {
+        PathfindingResult::Success(pathfinding_result_success) => Some((
+            &pathfinding_result_success.path.track_section_ranges,
+            &pathfinding_result_success.path_item_positions,
+        )),
+        PathfindingResult::Failure(_) => None,
+    }
+}
+
+/// Extract report train from simulation result
+/// Returns None if simulation failed
+pub fn extract_simulation_report_train(simulation: &simulation::Response) -> Option<&ReportTrain> {
+    match simulation {
+        simulation::Response::Success(SimulationResponseSuccess { final_output, .. }) => {
+            Some(&final_output.report_train)
+        }
+        _ => None,
+    }
+}
+
+/// Prepared data for computing track occupancy
+pub struct TrackOccupancyData<'a> {
+    pub path_projection: PathProjection<'a>,
+    pub path_item_positions: &'a [u64],
+    pub report_train: &'a ReportTrain,
+    pub path_item_id: Option<&'a str>,
+}
+
+/// Prepare all data needed for track occupancy computation
+/// This extracts pathfinding data, simulation data, and finds the matching path item
+pub fn prepare_track_occupancy_data<'a>(
```

Comment:
Feels like a "catch-all" structure (as the `-Data` suffix suggests lacking a more precise description) even though it doesn't "catch all". Here's my suggestion:

* remove the `report_train`
* rename the structure to `PathParameters` (we're already in `mod track_occupancy` and this structure should be private)
* extract the `report_train` directly in `find_track_occupancy_for_operational_point`
* use `PathParameters` in all functions instead of unwrapping it directly in `find_track_occupancy_for_operational_point`